### ISSUE TITLE
Correctly parse but reject certs for EC keys with implicitCurve

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#12895]
[RFC 5480 section 2.1.1](https://tools.ietf.org/html/rfc5480#section-2.1.1) requires that implicitCurve and specifiedCurve MUST NOT be used in PKIX.
Lacking a suitable example/test cert it is currently unclear if OpenSSL can parse ECC certs with implicitCurve.
While #12683 handles the specifiedCurve case, I suspect that rejection of both cases can be achieved simply by checking
```
EC_GROUP_get_curve_name(EC_KEY_get0_group(EVP_PKEY_get0_EC_KEY(pkey))) != NID_undef
```

This is a spin-off of the discussion in https://github.com/openssl/openssl/pull/12683#discussion_r486980401.

**Comment by t8m** [comment]
> 
> ```
> EC_GROUP_get_curve_name(EC_KEY_get0_group(EVP_PKEY_get0_EC_KEY(pkey))) != NID_undef
> ```

No, that would not work. The curve name is actually set accordingly when parsing a certificate whose explicit parameters happen to match one of the named curves.

**Comment by t8m** [comment]
Besides, how would you like to "correctly parse" the implicitCurve certs when you do not really know the parameters when parsing a single cert. Do you want to try to look up the CA cert somewhere? And if you mean something else by the "correctly parse" then please specify.

**Comment by DDvO** [comment]
> > ```
> > EC_GROUP_get_curve_name(EC_KEY_get0_group(EVP_PKEY_get0_EC_KEY(pkey))) != NID_undef
> > ```
> 
> No, that would not work. The curve name is actually set accordingly when parsing a certificate whose explicit parameters happen to match one of the named curves.

I meanwhile found where this conversion is done, in `ec_lib.c`:
```
static EC_GROUP *ec_group_explicit_to_named(const EC_GROUP *group, OPENSSL_CTX *libctx, const char *propq, BN_CTX *ctx)
{
...
    if ((curve_name_nid = ec_curve_nid_from_params(dup, ctx)) != NID_undef) {
        /*
         * The input explicit parameters successfully matched one of the
         * built-in curves: often for built-in curves we have specialized
         * methods with better performance and hardening.
         *
         * In this case we replace the `EC_GROUP` created through explicit
         * parameters with one created from a named group.
         */
```
As I already wrote in https://github.com/openssl/openssl/pull/12683#discussion_r487542782,
if an explicit curve happens to match a named (predefined) curve, which means that they are semantically equal,
I'd say this does not need to be flagged as wrong because the essence of the RFC requirement is that 
'custom' curves must not be used (due to security considerations).

**Comment by DDvO** [comment]
> Besides, how would you like to "correctly parse" the implicitCurve certs when you do not really know the parameters when parsing a single cert. Do you want to try to look up the CA cert somewhere? And if you mean something else by the "correctly parse" then please specify.

@t8m, look what I've just digged out - the ancient function `X509_get_pubkey_parameters()` in `x509_vfy.c`
available from the very beginning of OpenSSL:
```
commit 58964a492275ca9a59a0cd9c8155cb2491b4b909
Author: Ralf S. Engelschall <rse@openssl.org>
Date:   Mon Dec 21 10:56:39 1998 +0000

    Import of old SSLeay release: SSLeay 0.9.0b
```
which is undocumented but apparently fills in implicit public key parameters!

**Comment by DDvO** [comment]
> And if you mean something else by the "correctly parse" then please specify.

Well, I was uncertain on this, but meanwhile I found another interesting piece of code in `x_pubkey.c` that essentially answers this:
```
EVP_PKEY *X509_PUBKEY_get0(const X509_PUBKEY *key)
{
..
    if (key->pkey != NULL)
        return key->pkey;

    /*
     * When the key ASN.1 is initially parsed an attempt is made to
     * decode the public key and cache the EVP_PKEY structure. If this
     * operation fails the cached value will be NULL. Parsing continues
     * to allow parsing of unknown key types or unsupported forms.
     * We repeat the decode operation so the appropriate errors are left
     * in the queue.
     */
```
And in fact I was able to produce a test cert file with an implicitCurve:
```
123  11:       SEQUENCE {
125   7:         OBJECT IDENTIFIER ecPublicKey (1 2 840 10045 2 1)
134   0:         NULL
       :         }
```
and it does get parsed with the following entries in the error queue being ignored deliberately:
```
openssl x509 -noout -text -in test/certs/p384-server-cert-implicitCurve.pem
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: CN = ECDSA P-384 root
        Validity
            Not Before: Feb 18 18:49:54 2017 GMT
            Not After : Feb 19 18:49:54 2117 GMT
        Subject: CN = P-384 Server Cert
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
            Unable to load Public Key
140582513763456:error:100DC08E:elliptic curve routines:eckey_type2param:decode error:../crypto/ec/ec_ameth.c:124:
140582513763456:error:100D7010:elliptic curve routines:eckey_pub_decode:EC lib:../crypto/ec/ec_ameth.c:151:
140582513763456:error:0B09407D:x509 certificate routines:x509_pubkey_decode:public key decode error:../crypto/x509/x_pubkey.c:125:
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                98:F9:F7:EA:00:A7:81:A6:24:25:E1:BE:A1:8D:43:97:B8:16:C8:96
            X509v3 Authority Key Identifier: 
                keyid:26:D0:8F:1D:7B:5F:DC:1E:7F:41:80:7D:63:CA:1C:75:81:D6:85:69

            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Subject Alternative Name: 
                DNS:P-384 Server Cert
    Signature Algorithm: ecdsa-with-SHA256
         30:65:02:31:00:c3:d6:ec:e6:0c:cc:40:9a:4c:a8:90:89:b3:
         5f:88:56:e9:fd:87:c6:df:e2:64:98:53:e8:2d:79:cc:28:d8:
         7e:85:9c:05:d3:ff:12:21:79:9b:61:4c:3d:4b:f1:ab:ed:02:
         30:09:8a:3a:77:5c:7c:ec:10:ec:ea:1a:11:4c:92:c7:6a:b5:
         d2:25:e0:ae:41:38:d8:7b:13:bc:51:11:f0:55:0d:ee:b7:a9:
         39:49:c3:26:10:bd:50:a2:56:ec:8e:40:f7
```

**Comment by t8m** [comment]
> As I already wrote in [#12683 (comment)](https://github.com/openssl/openssl/pull/12683#discussion_r487542782),
> if an explicit curve happens to match a named (predefined) curve, which means that they are semantically equal,
> I'd say this does not need to be flagged as wrong because the essence of the RFC requirement is that
> 'custom' curves must not be used (due to security considerations).

I disagree with that. I do not see anything like that could be inferred from the RFC wording. There is simply: ` -- implicitCurve and specifiedCurve MUST NOT be used in PKIX`. I do not see how you could interpret it in the above way.

**Comment by t8m** [comment]
OK, so a cert with implicitCurve is parsed correctly, but errors are left on the error stack. But the certificate won't be usable with the current chain verification code anyway.

**Comment by DDvO** [comment]
> OK, so a cert with implicitCurve is parsed correctly, but errors are left on the error stack.

Right, but the errors were misleading too bulky. This is fixed by #13003.

> But the certificate won't be usable with the current chain verification code anyway.

This is also fixed by #13003, using a new `-inherit_pkparam` option etc.

**Comment by kaduk** [comment]
> I do not see anything like that could be inferred from the RFC wording. There is simply: -- implicitCurve and specifiedCurve MUST NOT be used in PKIX. I do not see how you could interpret it in the above way.

I agree with @t8m -- the meaning of the text is clear, even if it is arguably overzealous in terms of what is needed to provide the security benefits.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
